tests/core/data/test_tensordense.py

```diff
@@ -0,0 +1,234 @@
+        pytest.param(_valid_list(), {'shape': ()}, ValueError,
+                     id="numpy-shape 0 tuple"),
+        pytest.param(_valid_list(), {'shape': (1,)}, ValueError,
+                     id="numpy-shape 1 tuple"),
+        pytest.param(_valid_list(), {'shape': (None, None)}, ValueError,
+                     id="numpy-shape None tuple"),
+        pytest.param(_valid_list(), {'shape': [2, 2]}, ValueError,
+                     id="numpy-shape list"),
+        pytest.param(_valid_list(), {'shape': (1, 2, 3)}, ValueError,
+                     id="numpy-shape 3 tuple"),
+        pytest.param(_valid_list(), {'shape': (-1, 1)}, ValueError,
+                     id="numpy-negative shape"),
+        pytest.param(_valid_list(), {'shape': (-4, -4)}, ValueError,
+                     id="numpy-both negative shape"),
+        pytest.param(_valid_list(), {'shape': (1213, 1217)}, ValueError,
+                     id="numpy-different shape"),
+        pytest.param(_valid_tensor(), {'shape': ()}, ValueError,
+                     id="numpy-shape 0 tuple"),
+        pytest.param(_valid_tensor(), {'shape': (1,)}, ValueError,
+                     id="numpy-shape 1 tuple"),
+        pytest.param(_valid_tensor(), {'shape': (None, None)}, ValueError,
+                     id="numpy-shape None tuple"),
+        pytest.param(_valid_tensor(), {'shape': [2, 2]}, ValueError,
+                     id="numpy-shape list"),
+        pytest.param(_valid_tensor(), {'shape': (1, 2, 3)}, ValueError,
+                     id="numpy-shape 3 tuple"),
+        pytest.param(_valid_tensor(), {'shape': (-1, 1)}, ValueError,
+                     id="numpy-negative shape"),
+        pytest.param(_valid_tensor(), {'shape': (-4, -4)}, ValueError,
+                     id="numpy-both negative shape"),
+        pytest.param(_valid_tensor(), {'shape': (1213, 1217)}, ValueError,
+                     id="numpy-different shape"),
+    ])
+    def test_init_from_wrong_input(self, arg, kwargs, error):
+        """
+        Test that the __init__ method raises a suitable error when passed
+        incorrectly formatted inputs.
+
+        This test also serves as a *partial* check that Dense safely handles
+        deallocation in the presence of exceptions in its __init__ method.  If
+        the tests segfault, it's quite likely that the memory management isn't
+        being done correctly in the hand-off us setting our data buffers up and
+        marking the numpy actually owns the data.
+        """
+        with pytest.raises(error):
+            DenseTensor(arg, **kwargs)
+
+    def test_copy_returns_a_correct_copy(self, data_tensor_dense):
+        """
+        Test that the copy() method produces an actual copy, and that the
+        result represents the same matrix.
+        """
+        original = data_tensor_dense
+        copy = data_tensor_dense.copy()
+        assert original is not copy
+        assert np.all(original._tf == copy._tf)
```

Comment:
Is numpy the right tool to test tensor's equality?

Reply:
I have not found anything similar in TensorFlow.